src/morphology.rs

```diff
@@ -576,33 +588,76 @@ impl Mask {
     /// # }
     /// ```
     pub fn disk(radius: u8) -> Self {
-        let range = -(radius as i16)..=(radius as i16);
-        Self {
-            elements: range
+        let radius_squared = u32::from(radius).pow(2);
+        let half_widths_per_height = std::iter::successors(
+            Some((-i16::from(radius), 0u8)),
+            |(last_height, last_half_width)| {
+                if *last_height == i16::from(radius) {
+                    return None;
+                };
+                let next_height = last_height + 1;
+                let height_squared = (u32::from(next_height.unsigned_abs())).pow(2);
+                let next_half_width = if next_height <= 0 {
+                    // upper part of the circle => increasing width
+                    (u32::from(*last_half_width)..)
+                        .find(|x| (x + 1).pow(2) + height_squared > radius_squared)?
+                } else {
+                    // lower part of the circle => decreasing width
+                    (0..=u32::from(*last_half_width))
+                        .rev()
+                        .find(|&x| x.pow(2) + height_squared <= radius_squared)?
+                } as u8; // next_half_width always fits in a u8 because he radius itself is a u8
+                Some((next_height, next_half_width))
+            },
+        );
+        let mut elements = Vec::with_capacity(
+            half_widths_per_height
                 .clone()
-                .cartesian_product(range)
-                .filter(|(x, y)| {
-                    (x.unsigned_abs() as u32).pow(2) + (y.unsigned_abs() as u32).pow(2)
-                        <= (radius as u32).pow(2)
-                })
-                .collect(),
-        }
+                .map(|(_, half_width)| 2 * usize::from(half_width) + 1)
+                .sum(),
+        );
+        elements.extend(half_widths_per_height.flat_map(|(y, half_width)| {
+            ((-i16::from(half_width))..=i16::from(half_width)).map(move |x| Point::new(x, y))
+        }));
+        Self::new(elements)
     }
 
-    fn apply<'a, 'b: 'a, 'c: 'a>(
-        &'c self,
-        image: &'b GrayImage,
-        x: u32,
-        y: u32,
-    ) -> impl Iterator<Item = &'a Luma<u8>> {
-        self.elements
-            .iter()
-            .map(move |(i, j)| (x as i64 + *i as i64, y as i64 + *j as i64))
-            .filter(move |(i, j)| {
-                0 <= *i && *i < image.width() as i64 && 0 <= *j && *j < image.height() as i64
-            })
-            .map(move |(i, j)| image.get_pixel(i as u32, j as u32))
+    /// all the slices are garanteed to be non empty
+    fn lines<'a>(
+        &'a self,
+    ) -> GroupBy<i16, Iter<Point<i16>>, impl FnMut(&&'a Point<i16>) -> i16 + '_> {
+        self.elements.iter().group_by(|p: &&'a Point<i16>| p.y)
```

Comment:
This seems to assume that the elements are sorted into y_groups. Should be enforce this ordering in `Mask::new()`? Or is this just for performance reasons in case points are listed in y-groups which might always be the case for the various constructors?